### PR TITLE
Allow Hybrid Composition fallback for Android platform views

### PR DESCRIPTION
**Pull request by stuartmorgan**
Allows for the new Android platform view mode introduced in Flutter 3.0 (tentatively called Texture Layer Hybrid Composition) to fall back—when explicitly requested—to Hybrid Composition rather than Virtual Display when TLHC mode is not supported. This allows us to avoid the issue introduced in Flutter 3.0 where a plugin that previously used HC will now use VD in some cases, which would not be the plugin author's intent since they had explicitly chosen the initialization mode that was for HC in Flutter 2.x.

In order to support this, the internals of creation of each mode have been extracted to helper methods, to allow for delegation, and the use of a mutable wrapper context is now unconditional as we now no longer know if we will need it until after the context has been provided to the plugin's factory method; we just don't ever set the underlying context to anything else in the cases where it's not needed.

This is the engine portion of https://github.com/flutter/flutter/issues/107313
Framework portion: https://github.com/flutter/flutter/pull/109161

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by stuartmorgan** [comment]
@bparrishMines @dnfield This is a WIP since I still need to add tests, but I wanted to post it to get initial feedback on the approach.

File: testing/scenario_app/lib/src/platform_view.dart [hunk 1]
```diff
@@ -28,6 +28,13 @@ List<int> _to64(num value) {
   return temp;
 }
 
+List<int> _encodeString(String value) {
```
Comment (stuartmorgan) [inline testing/scenario_app/lib/src/platform_view.dart:31]: This was just a drive-by improvement since I accidentally mismatched the strings in the `.length` and `utf8.encode` lines when I added a new string, so I thought I'd prevent the whole class of mistake once I realized what I'd done wrong.

**Comment by stuartmorgan** [comment]
The new golden tests I added have somewhat odd visual output where the top of the platform view square is white, instead of pale grey like most of the other outputs from this suite of tests. I'm not sure why (and whether it's a problem with the manual composition these tests do, or a real issue with HC + `SurfaceView`), but since it's the same for both the "HC with SurfaceView" and "fallback from TLHC to HC with SurfaceView" tests I added, I decided to just go ahead with landing them since it means it's not specific to the new fallback logic.

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 11.
View them at https://flutter-engine-gold.skia.org/cl/github/35233

**Review by bparrishMines — APPROVED** [review]
LGTM

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 12.
View them at https://flutter-engine-gold.skia.org/cl/github/35233

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 13.
View them at https://flutter-engine-gold.skia.org/cl/github/35233
